src/main/java/org/folio/circulation/domain/OverdueFineCalculatorService.java

```diff
@@ -0,0 +1,201 @@
+    OverdueFinePolicyRepository overdueFinePolicyRepository,
+    AccountRepository accountRepository,
+    ItemRepository itemRepository,
+    FeeFineOwnerRepository feeFineOwnerRepository,
+    FeeFineRepository feeFineRepository,
+    OverduePeriodCalculatorService overduePeriodCalculatorService) {
+    this.overdueFinePolicyRepository = overdueFinePolicyRepository;
+    this.accountRepository = accountRepository;
+    this.itemRepository = itemRepository;
+    this.feeFineOwnerRepository = feeFineOwnerRepository;
+    this.feeFineRepository = feeFineRepository;
+    this.overduePeriodCalculatorService = overduePeriodCalculatorService;
+  }
+
+  private OverdueFineCalculatorService(Clients clients) {
+    this(new OverdueFinePolicyRepository(clients),
+      new AccountRepository(clients),
+      new ItemRepository(clients, true, false, false),
+      new FeeFineOwnerRepository(clients),
+      new FeeFineRepository(clients),
+      new OverduePeriodCalculatorService(new CalendarRepository(clients)));
+  }
+
+  public CompletableFuture<Result<CheckInProcessRecords>> calculateOverdueFine(
+    CheckInProcessRecords records) {
+    Loan loan = records.getLoan();
+    if (loan == null) {
+      return completedFuture(of(() -> records));
+    }
+
+    return calculateOverdueFine(loan).thenApply(mapResult(v -> records));
+  }
+
+  public CompletableFuture<Result<LoanAndRelatedRecords>> calculateOverdueFine(
+    LoanAndRelatedRecords records) {
+    Loan loan = records.getLoan();
+    if (loan == null) {
+      return completedFuture(of(() -> records));
+    }
+    return calculateOverdueFine(loan).thenApply(mapResult(v -> records));
+  }
+
+  private CompletableFuture<Result<Account>> calculateOverdueFine(Loan loan) {
+    return completedFuture(succeeded(loan))
+      .thenComposeAsync(overdueFinePolicyRepository::findOverdueFinePolicyForLoan)
+      .thenCompose(r -> r.after(l -> getOverdueMinutes(l)
+          .thenCompose(mr -> mr.after(minutes -> calculateOverdueFine(l, minutes)))
+          .thenCompose(fr -> fr.after(fine -> this.createFeeFineRecord(l, fine)))
+      ));
+  }
+
+  private CompletableFuture<Result<Integer>> getOverdueMinutes(Loan loan) {
+    DateTime systemTime = loan.getReturnDate();
+    if (systemTime == null) {
+      systemTime = DateTime.now(DateTimeZone.UTC);
+    }
+    return overduePeriodCalculatorService.getMinutes(loan, systemTime);
+  }
+
+  private CompletableFuture<Result<Double>> calculateOverdueFine(Loan loan, Integer overdueMinutes) {
+    double maxFine = loan.wasDueDateChangedByRecall() ?
+      loan.getOverdueFinePolicy().getMaxOverdueRecallFine() :
+      loan.getOverdueFinePolicy().getMaxOverdueFine();
+    OverdueFineInterval interval = loan.getOverdueFinePolicy().getOverdueFineInterval();
+    double overdueFine = loan.getOverdueFinePolicy().getOverdueFine() *
+      Math.ceil(overdueMinutes.doubleValue()/MINUTES_IN_INTERVAL.get(interval).doubleValue());
+    if (maxFine > 0) {
```

Comment:
Personally, I would use some whitespace to split up these statements, particularly the ternary operator and the variable below it.

Reply:
Done